File: lt1022.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x){ val = x;}
}


class Solution {
    public int sumRootToLeaf(TreeNode root) {
        sumRootToLeaf(root, 0);
        return res; 
    }
    
    public void sumRootToLeaf(TreeNode root, int val){
        if(root == null){
            return;
        }
        
        int nVal = (val * 2 + root.val) % mod;
        
        if(root.left == null && root.right ==null){
            res = (res + nVal) % mod;
        }
        
        sumRootToLeaf(root.left,  nVal);
        sumRootToLeaf(root.right, nVal);
    }
    
    private int res = 0;
    private final int mod = (int)Math.pow(10, 9) + 7;
}

File: code/lt1022.cpp
// https://leetcode-cn.com/problems/sum-of-root-to-leaf-binary-numbers/
// sum of root to leaf binary numbers
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        int sum = 0;
        helper(root, 0, sum);
        return sum;
    }
    
    void helper(TreeNode* node, int prefix, int& sum){
        if(node == nullptr)
                return;
        
        int cur = prefix * 2 + node->val;
        
        if(node->left == nullptr && node->right == nullptr){
            sum += cur;
            return;
        }
        
        helper(node->left,  cur, sum);
        helper(node->right, cur, sum);
    }
    
};